### DISCUSSION TITLE
Keep trailing zero when serializing BigDecimal to JSON

### DISCUSSION THREAD

**Discussion by caodangtinh** [FasterXML/jackson-databind#4667] [General]
### Describe your Issue

Is there any way to keep the trailing zero value when serializing BigDecimal to JSON? I'd expect something like
```
"version": 2.0
```
But I always get
```
"version": 2
```
I have already tried with
```
@JsonFormat(shape = JsonFormat.Shape.NUMBER, pattern = "#.#")
```
And a customer, JsonSerializer
```
public class BigDecimalStringSerializer extends JsonSerializer<BigDecimal> {

    @Override
    public void serialize(BigDecimal value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
        gen.writeNumber(value.toPlainString());
    }
}
```
But it is still produce the same output.

**Comment by pjfanning** [comment]
Issues are for bugs. So far, this appears just to be a usage question.

Have you tried pattern = "#.0"? I don't know for definite but BigDecimal formatting should be handled by this class
https://docs.oracle.com/javase/8/docs/api/java/text/DecimalFormat.html

And its Javadocs describe how the pattern values like '#' and '0' work.

**Reply by caodangtinh** [reply]
@pjfanning , I have tried with
```
@JsonFormat(shape = JsonFormat.Shape.NUMBER, pattern = "#.0")
```
As well as the DecimalFormat
```
public class BigDecimalStringSerializer extends JsonSerializer<BigDecimal> {
    @Override
    public void serialize(BigDecimal value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
        var df = new DecimalFormat();
        df.setMinimumFractionDigits(value.scale());
        df.setMaximumFractionDigits(value.scale());
        df.setGroupingUsed(false);
        gen.writeNumber(df.format(value));
    }
}
```
However, it still got the same result, trailing zero was removed.
I'm using Jackson version 2.17.2

**Reply by pjfanning** [reply]
I tried using the JsonFormat annotation and the pattern does seem not to work.
I used format shape String but also tried Number and in both cases the pattern had no effect.
My example is in https://github.com/FasterXML/jackson-databind/compare/2.18...pjfanning:jackson-databind:number-serialize-pattern

@cowtowncoder Any suggestions? I can't find any similar serialization tests where numbers are formatted using the JsonFormat annotation.

**Reply by pjfanning** [reply]
https://github.com/FasterXML/jackson-databind/pull/4671 seems to show that custom serializers can be made to work.

**Reply by cowtowncoder** [reply]
Correct: `@JsonFormat` does not change Number serialization at this point.

But custom serializer should work, esp. if using String-taking method, `writeNumber(String numStr)`

**Reply by yihtserns** [reply]
> ...
> df.setMinimumFractionDigits(value.scale());
> ...
> However, it still got the same result, trailing zero was removed.

Likely because `value.scale()` returned `0`.  You should've just do `df.setMinimumFractionDigits(1)` and it would've worked.

**Comment by yihtserns** [comment]
One way to serialize with trailing zero without needing `@JsonFormat` nor custom `JsonSerializer`  is to create a `BigDecimal` with trailing zero, e.g.:
- `BigDecimal.valueOf(2.0)`
- `new BigDecimal("2.0")`

**Reply by caodangtinh** [reply]
Nice, it does work
